source/agora/consensus/EnrollmentManager.d

```diff
@@ -682,6 +642,62 @@ public class EnrollmentManager
         }
     }
 
+    /***************************************************************************
+
+        Generate pre-images
+
+        Params:
+            current_height = block height used to determine which range of
+                pre-images will be generated.
+
+    ***************************************************************************/
+
+    private void generatePreimages(ulong current_height) @safe nothrow
+    {
+        // This determines which range of preimages must be generated.
+        ulong start_height =
+            ((current_height / ValidatorSet.ValidatorCycle) + 2) *
+                ValidatorSet.ValidatorCycle;
+        ulong bulk_index = start_height / AllCountPrimages;
+        const ulong MaxHeight = (bulk_index + 1) * AllCountPrimages;
+
+        // Clear if recreating pre-images is needed
+        () @trusted {
+            if (this.preimage_rounds.length > 0 &&
+                sort(this.preimage_rounds.keys)[$-1] < start_height)
```

Comment:
`std.algorithm.max` and `byKey`

Reply:
> `std.algorithm.max` and `byKey`

Can you show me how to use the two functions for that? I can't find an exampe or what the byKey returns.

Reply:
Is the code below what you suggested?
```
ulong max_height = 0;
foreach (key; this.preimage_rounds.byKey)
    max_height = max(key, max_height);
```

Reply:
Almost. It's actually called https://dlang.org/phobos/std_algorithm_searching.html#.maxElement
So:
```D
auto max_height = this.preimage_rounds.byKey.maxElement();
```